File: Array/Easy/53-Maximum-Subarray/solution_prefixSum.cs
public class Solution {
    public int MaxSubArray(int[] nums) {
        if(nums == null || nums.Length == 0) {
            throw new ArgumentException("The argument provided is invalid.");
        }
        int max = nums[0];
        int sum = 0;
        foreach(int num in nums) {
            sum += num;
            max = Math.Max(sum, max);
            sum = sum < 0 ? 0 : sum;  // not contributing to the onward subarray sum, set to 0
        }
        return max;
    }
}

File: Array/Easy/Lint44-Minimum-Subarray/solution_prefixSum.cpp
class Solution {
public:
    /*
     * @param nums: a list of integers
     * @return: A integer indicate the sum of minimum subarray
     */
    int minSubArray(vector<int> &nums) {
        // prefix sum => minSum = prefixSum[i + 1] - max{prefixSum[0], ... , prefixSum[i]}
        // tc:O(n); sc:O(1)
        if(nums.size() == 0) {
            return 0;
        }
        int sum = 0;
        int maxSum = 0;
        int res = nums[0];
        for(int num : nums) {
            sum += num;
            res = res > sum - maxSum ? sum - maxSum : res;
            maxSum = maxSum >= sum ? maxSum : sum;
        }
        return res;
    }
};